extensions.html

```diff
@@ -445,16 +507,168 @@ <h2>
           The current pose of the gamepad, if supported by the hardware.
           Includes position, orientation, velocity, and acceleration. If the
           hardware cannot supply any pose values, MUST be set to
-          <code>null</code>.
+          `null`.
         </dd>
         <dt>
           <dfn>touchEvents</dfn>
         </dt>
         <dd>
-          A list of touch events generated from all touch surfaces.
-          <code>null</code> if the device does not support touch events.
+          <p>A list of touch events generated from all touch surfaces.
+          If the device does not support touch events, MUST be set to `null`.</p>
+          <p>
+            The {{Gamepad/touchEvents}} getter steps are:
+          </p>
+          <ol>
+            <li>If [=this=].{{Gamepad/[[touchEvents]]}} not `null` and not empty,
+              return [=this=].{{Gamepad/[[touchEvents]]}}.
+            </li>
+            <li>Otherwise return `null`.
+            </li>
+          </ol>
+
         </dd>
       </dl>
+      <section>
+        <h3>
+          Receiving inputs
+        </h3>
+        <p>
+          This section supplements the <a href="https://w3c.github.io/gamepad/#receiving-inputs">
+          Receiving inputs</a> section of the main 
+          <a href="https://w3c.github.io/gamepad/">Gamepad specification</a>.
+        </p>
+        <p>
+          In addtion to the steps defined in the main 
+          <a href="https://w3c.github.io/gamepad/">Gamepad specification</a>.
+          When the user agent modifies the list of |active touch points| for a 
+          touch surface by adding touch points, removing touch points, 
+          or updating the values of existing touch points,
+          <dfn data-lt="updating touchEvents">update touchEvents</dfn>
+          by running the following steps:
+        </p>
+        <ol>
+          <li>Let |surfaceId:unsigned long| be 0.
+          </li>
+          <li>Let |touchSurfaces:list| be an [=ordered set=] of touch surfaces
```

Comment:
In this algorithm we describe `touchSurfaces`, `touchInputs`, `touchData` but there's no instructions on how to construct those objects.

We want to emphasize the assumptions we're making about the inputs but we also don't want to specify more than we have to. I think we can modify the algorithm to be more specific about the assumptions we're making while being less specific about how the data is presented. For example, if we specify that a gamepad's touch surfaces need to have some consistent ordering then we can iterate over them without creating lists:

* Let |surfaceId| be 0.
* Repeat these steps for each [=touch surface=] on |gamepad| in [=touch surface enumeration order=]:
  * Let |surfaceDimensions| be `null`.
  * If the [=touch surface=] exposes maximum dimensions in device units:
    * Set |surfaceDimensions| to a [=new=] {{Array}} with length 2.
    * Set |surfaceDimensions|[0] to the maximum X dimension of the current [=touch surface=] in device units.
    * Set |surfaceDimensions|[1] to the maximum Y dimension of the current [=touch surface=] in device units.
  * Repeat these steps for each [=active touch point=] on the current [=touch surface=]:
    * Let |x:float| be the X coordinate for the current [=active touch point=] in device units.
    * Let |y:float| be the Y coordinate for the current [=active touch point=] in device units.
    * Let |position| be a [=new=] {{Array}} with length 2.
    * If |surfaceDimensions| is `null`:
      * Set |position|[0] to ???.
      * Set |position|[1] to ???.
    * Otherwise:
      * Set |position|[0] to a value calculated from |x| and |surfaceDimensions|[0] and normalized to [-1,1].
      * Set |position|[1] to a value calculated from |y| and |surfaceDimensions|[1] and normalized to [-1,1].
    * Let |touchEvent| be a [=new=] {{GamepadTouch}}:
      * Initialize |touchEvent|.{{GamepadTouch/surfaceId}} to |surfaceId|.
      * Initialize |touchEvent|.{{GamepadTouch/position}} to |position|.
      * Initialize |touchEvent|.{{GamepadTouch/surfaceDimensions}} to |surfaceDimensions|.
      * [=list/Append=] |touchEvent| to |gamepad|.{{Gamepad/[[touchEvents]]}}.
  * Increment |surfaceId|.

Reply:
> Normalizing touchId in Dualshock4Controller won't work because a device::Gamepad isn't 1-to-1 with a blink::Gamepad. Most of the time, "internal slot" should be read as "1-to-1 with blink::Gamepad". The only time we can attach internal slots to device::Gamepad is if all tabs should see the same value. This isn't true for touchId and nextTouchId since they may differ between tabs even for the same gamepad.

I went back and double checked and what I have observed is that the lifetime of a blink::Gamepad is very short maybe each render, so will not work for touchId and nextTouchId. This is probably not the forum for this discussion (I need to figure out how to start it over on the chromium site).
This does create an issue with implementing the specification since the lifetime of the device::Gamepad is not correct either as you point out.

> We want to emphasize the assumptions we're making about the inputs but we also don't want to specify more than we have to. I think we can modify the algorithm to be more specific about the assumptions we're making while being less specific about how the data is presented. For example, if we specify that a gamepad's touch surfaces need to have some consistent ordering then we can iterate over them without creating lists:

Agreed, I will work on addressing this.

Reply:
Ah right, blink::Gamepad is short-lived. The specification says it should behave like a live object but the Chromium implementation hasn't been updated to match that behavior yet. (Relevant spec issue: https://github.com/w3c/gamepad/issues/8)

We've worked around this by attaching the internal state to blink::NavigatorGamepad instead of blink::Gamepad and creating a Gamepad::Client interface to provide access to this internal state so it's persisted even when the blink::Gamepad is destroyed and recreated.

> This is probably not the forum for this discussion (I need to figure out how to start it over on the chromium site).

Let's use this crbug to discuss the Chromium implementation: [1143942: Gamepad Multitouch](https://crbug.com/1143942)

Reply:
@nondebug I have restructured the Receiving inputs section and added a Glossary for terms. Several updates to your other comments we made a bit back, should I just resolve those comments? 
There is a suggestion at the top "Still in progress? Convert to draft", does that make since?